src/flint/types/fmpz_mpoly.pyx

```diff
@@ -135,362 +191,893 @@ cdef class fmpz_mpoly(flint_mpoly):
+                return self
+        return NotImplemented
+
     def __mul__(self, other):
         cdef fmpz_mpoly res
-        if typecheck(self, fmpz_mpoly):
-            if typecheck(other, fmpz_mpoly):
-                if (<fmpz_mpoly>self).ctx is not (<fmpz_mpoly>other).ctx:
-                    ctx, (self, other) = coerce_fmpz_mpolys(self, other)
-                res = fmpz_mpoly.__new__(fmpz_mpoly)
-                res.ctx = (<fmpz_mpoly>self).ctx
-                fmpz_mpoly_init(res.val, res.ctx.val)
-                res._init = True
-                fmpz_mpoly_mul(res.val, (<fmpz_mpoly>self).val, (<fmpz_mpoly>other).val, res.ctx.val)
-                return res
-            else:
-                other = any_as_fmpz(other)
-                if other is not NotImplemented:
-                    res = fmpz_mpoly.__new__(fmpz_mpoly)
-                    res.ctx = (<fmpz_mpoly>self).ctx
-                    fmpz_mpoly_init(res.val, res.ctx.val)
-                    res._init = True
-                    fmpz_mpoly_scalar_mul_fmpz(res.val, (<fmpz_mpoly>self).val, (<fmpz>other).val, res.ctx.val)
-                    return res
+        if typecheck(other, fmpz_mpoly):
+            if (<fmpz_mpoly>self).ctx is not (<fmpz_mpoly>other).ctx:
+                raise IncompatibleContextError(f"{(<fmpz_mpoly>self).ctx} is not {(<fmpz_mpoly>other).ctx}")
+            res = create_fmpz_mpoly(self.ctx)
+            fmpz_mpoly_mul(res.val, (<fmpz_mpoly>self).val, (<fmpz_mpoly>other).val, res.ctx.val)
+            return res
         else:
-            self = any_as_fmpz(self)
-            if self is not NotImplemented:
-                res = fmpz_mpoly.__new__(fmpz_mpoly)
-                res.ctx = (<fmpz_mpoly>other).ctx
-                fmpz_mpoly_init(res.val, res.ctx.val)
-                res._init = True
-                fmpz_mpoly_scalar_mul_fmpz(res.val, (<fmpz_mpoly>other).val, (<fmpz>self).val, res.ctx.val)
+            other = any_as_fmpz(other)
+            if other is not NotImplemented:
+                res = create_fmpz_mpoly(self.ctx)
+                fmpz_mpoly_scalar_mul_fmpz(res.val, (<fmpz_mpoly>self).val, (<fmpz>other).val, res.ctx.val)
                 return res
         return NotImplemented
 
+    def __rmul__(self, other):
+        cdef fmpz_mpoly res
+        other = any_as_fmpz(other)
+        if other is not NotImplemented:
+            res = create_fmpz_mpoly(self.ctx)
+            fmpz_mpoly_scalar_mul_fmpz(res.val, (<fmpz_mpoly>self).val, (<fmpz>other).val, res.ctx.val)
+            return res
+        return NotImplemented
+
+    def __imul__(self, other):
+        if typecheck(other, fmpz_mpoly):
+            if (<fmpz_mpoly>self).ctx is not (<fmpz_mpoly>other).ctx:
+                raise IncompatibleContextError(f"{(<fmpz_mpoly>self).ctx} is not {(<fmpz_mpoly>other).ctx}")
+            fmpz_mpoly_mul((<fmpz_mpoly>self).val, (<fmpz_mpoly>self).val, (<fmpz_mpoly>other).val, self.ctx.val)
+            return self
```

Comment:
This is different from how univariate polynomials work in python-flint:
```
In [85]: p = fmpz_poly([1, 2, 3])

In [86]: p
Out[86]: 3*x^2 + 2*x + 1

In [87]: p2 = p

In [88]: p *= p

In [89]: p
Out[89]: 9*x^4 + 12*x^3 + 10*x^2 + 4*x + 1

In [90]: p2
Out[90]: 3*x^2 + 2*x + 1

In [91]: ctx = fmpq_mpoly_ctx(2, Ordering.lex, ['x','y'])

In [93]: p3 = fmpq_mpoly('2*x^2 + 2*x + 1', ctx)

In [94]: p4 = p3

In [95]: p3 *= p3

In [96]: p3
Out[96]: 4*x^4 + 8*x^3 + 8*x^2 + 4*x + 1

In [97]: p4
Out[97]: 4*x^4 + 8*x^3 + 8*x^2 + 4*x + 1
```
There is an unresolved tension here about whether polynomials should behave like mutable or immutable objects. I imagine that having `p3 *= p3` cause an in-place mutation will trip someone up somewhere who doesn't realise that copies need to be made. Certainly in the context of SymPy the expectation is that such objects can be treated as immutable and therefore do not need to be copied.

Reply:
I can imagine that having `*=` perform an in-place mutation would mess something up somewhere when it is not expected.

This is a bit different from other more explicit mutations like `p[k] = v`.

Reply:
> This is a bit different from other more explicit mutations like p[k] = v.

I was going to comment about that. Immutability can be a pretty nice property but flint does support mutation and I think when it's an explicit "flint based" mutation it should be kept but the "mutations" imposed by Python operands should be avoided.

Reply:
I agree it would be nice to expose mutability in some way but overloading the ``__iadd__`` and ``__imul__`` operators will break generic Python code.

Reply:
Perhaps there could be explicit methods like `p1.imul(p2)` for in-place mutation.

There could also be a `p.hash()` method rather than `__hash__` and then a wrapper class that treats polynomials as immutable can do:
```python
def __hash__(self):
    return self._rep.hash()
```

Reply:
> Perhaps there could be explicit methods like p1.imul(p2) for in-place mutation.

Have just added explicit `iadd`, `isub`, and `imul` methods based off the previous in-place methods.